kong/client.go

```diff
@@ -280,18 +309,20 @@ func (c *Client) Status(ctx context.Context) (*Status, error) {
 	return &s, nil
 }
 
-// Root returns the response of GET request on root of
-// Admin API (GET /).
+// Root returns the response of GET request on root of Admin API (GET / or /kong with a workspace).
 func (c *Client) Root(ctx context.Context) (map[string]interface{}, error) {
-	req, err := c.NewRequest("GET", "/", nil, nil)
+	endpoint := "/"
+	if c.hasWorkspace() {
+		endpoint = "/kong"
+	}
+	req, err := c.NewRequest("GET", endpoint, nil, nil)
```

Comment:
There is a synchronization issue here that's not trivial to fix.

`NewRequest` internally makes a call to `baseURL` which locks the workspace for reading.
Before that, we're making a workspace reading operation to check whether the endpoint shall be `/` or `/kong`.
We cannot allow these two operations to happen between synchronization in between because the workspace can change after the first lock gets released and before the second gets acquired.
We cannot just put another big lock on top of these two because that can lead into a deadlock.

The best way out of this situation I can think of is to refactor `NewRequest` so that there's a subset thereof that doesn't perform locking internally, and use it here.

That would mean the following (not intentionally breaking) change to `NewRequest`:

```go
func (c *Client) NewRequestRaw(method, baseURL string, endpoint string, qs interface{},                                                                                                                                                                                                                      
  body interface{}) (*http.Request, error) {                                                                                                                                                                                                                                              
  if endpoint == "" {                                                                                                                                                                                                                                                                     
    return nil, errors.New("endpoint can't be nil")                                                                                                                                                                                                                                       
  }                                                                                                                                                                                                                                                                                       
  //body to be sent in JSON                                                                                                                                                                                                                                                               
  var buf []byte                                                                                                                                                                                                                                                                          
  if body != nil {                                                                                                                                                                                                                                                                        
    var err error                                                                                                                                                                                                                                                                         
    buf, err = json.Marshal(body)                                                                                                                                                                                                                                                         
    if err != nil {                                                                                                                                                                                                                                                                       
      return nil, err                                                                                                                                                                                                                                                                     
    }                                                                                                                                                                                                                                                                                     
  }                                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                          
  //Create a new request                                                                                                                                                                                                                                                                  
  req, err := http.NewRequest(method, baseURL+endpoint,                                                                                                                                                                                                                               
    bytes.NewBuffer(buf))                                                                                                                                                                                                                                                                 
  if err != nil {                                                                                                                                                                                                                                                                         
    return nil, err                                                                                                                                                                                                                                                                       
  }                                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                          
  // add body if needed                                                                                                                                                                                                                                                                   
  if body != nil {                                                                                                                                                                                                                                                                        
    req.Header.Add("Content-Type", "application/json")                                                                                                                                                                                                                                    
  }                                                                                                                                                                                                                                                                                       
                                                                                                                                                                                                                                                                                          
  // add query string if any                                                                                                                                                                                                                                                              
  if qs != nil {                                                                                                                                                                                                                                                                          
    values, err := query.Values(qs)                                                                                                                                                                                                                                                       
    if err != nil {                                                                                                                                                                                                                                                                       
      return nil, err                                                                                                                                                                                                                                                                     
    }                                                                                                                                                                                                                                                                                     
    req.URL.RawQuery = values.Encode()                                                                                                                                                                                                                                                    
  }                                                                                                                                                                                                                                                                                       
  return req, nil                                                                                                                                                                                                                                                                         
}                        

func (c *Client) NewRequest(method, endpoint string, qs interface{},                                                                                                                                                                                                                      
  body interface{}) (*http.Request, error) {                                                                                                                                                                                                                                              
  return c.NewRequestRaw(method, c.baseURL(c.GetWorkspace()), endpoint, qs, body)
}                        

```

And this function calling `NewRequestRaw`, using the workspace that was read only once.